src/burn/engine/manifest.cpp

```diff
@@ -134,3 +143,119 @@ extern "C" HRESULT ManifestLoadXmlFromBuffer(
     ReleaseObject(pixdDocument);
     return hr;
 }
+
+static HRESULT ParseContainersFromXml(
+    __in BURN_SECTION* pSection,
+    __in BURN_CONTAINERS* pContainers,
+    __in IXMLDOMNode* pixnBundle
+    )
+{
+    HRESULT hr = S_OK;
+    IXMLDOMNodeList* pixnNodes = NULL;
+    IXMLDOMNode* pixnNode = NULL;
+    DWORD cNodes = 0;
+    LPWSTR scz = NULL;
+
+    // select container nodes
+    hr = XmlSelectNodes(pixnBundle, L"Container", &pixnNodes);
+    ExitOnFailure(hr, "Failed to select container nodes.");
+
+    // get container node count
+    hr = pixnNodes->get_length((long*) &cNodes);
+    ExitOnFailure(hr, "Failed to get container node count.");
+
+    if (!cNodes)
+    {
+        ExitFunction();
+    }
+
+    // allocate memory for searches
+    pContainers->rgBoxes = (WIX_BOX*) MemAlloc(sizeof(WIX_BOX) * cNodes, TRUE);
+    ExitOnNull(pContainers->rgBoxes, hr, E_OUTOFMEMORY, "Failed to allocate memory for container structs.");
+
+    pContainers->cBoxes = cNodes;
+
+    // parse search elements
+    for (DWORD i = 0; i < cNodes; ++i)
+    {
+        WIX_BOX* pBox = &pContainers->rgBoxes[i];
+
+        hr = XmlNextElement(pixnNodes, &pixnNode, NULL);
+        ExitOnFailure(hr, "Failed to get next node.");
+
+        // @Id
+        hr = XmlGetAttributeEx(pixnNode, L"Id", &pBox->sczId);
+        ExitOnFailure(hr, "Failed to get @Id.");
+
+        // @Primary
+        hr = XmlGetYesNoAttribute(pixnNode, L"Primary", &pBox->fPrimary);
+        if (E_NOTFOUND != hr)
+        {
+            ExitOnFailure(hr, "Failed to get @Primary.");
+        }
+
+        // @Attached
+        hr = XmlGetYesNoAttribute(pixnNode, L"Attached", &pBox->fAttached);
+        if (E_NOTFOUND != hr || pBox->fPrimary) // if it is a primary container, it has to be attached
+        {
+            ExitOnFailure(hr, "Failed to get @Attached.");
+        }
+
+        // @AttachedIndex
+        hr = XmlGetAttributeNumber(pixnNode, L"AttachedIndex", &pBox->dwAttachedIndex);
+        if (E_NOTFOUND != hr || pBox->fAttached) // if it is an attached container it must have an index
+        {
+            ExitOnFailure(hr, "Failed to get @AttachedIndex.");
+        }
+
+        // @FilePath
```

Comment:
We need to incorporate this new code from clean room somehow: 

```
        // Attached containers are always found attached to the current process, so use the current proccess's
        // name instead of what may be in the manifest.
        if (pContainer->fAttached)
        {
            hr = PathForCurrentProcess(&scz, NULL);
            ExitOnFailure(hr, "Failed to get path to current process for attached container.");

            LPCWSTR wzFileName = PathFile(scz);

            hr = StrAllocString(&pContainer->sczFilePath, wzFileName, 0);
            ExitOnFailure(hr, "Failed to set attached container file path.");
        }
        else
        {
            // @FilePath
            hr = XmlGetAttributeEx(pixnNode, L"FilePath", &pContainer->sczFilePath);
            if (E_NOTFOUND != hr)
            {
                ExitOnFailure(hr, "Failed to get @FilePath.");
            }
        }
```